lymph/midline.py

```diff
@@ -375,9 +597,79 @@ def check_and_assign(self, new_params: np.ndarray):
+
+        This method mainly exists so that the checking and assigning of the
+        spread probs can be skipped.
+        """
+        stored_t_stages = set(self.ipsi.diagnose_matrices.keys())
+        provided_t_stages = set(self.ipsi.diag_time_dists.keys())
+        t_stages = list(stored_t_stages.intersection(provided_t_stages))
+        
+        max_t = self.diag_time_dists.max_t
+        state_probs_ex = np.zeros(shape=len(self.state_list), dtype=float)
+        state_probs_ex[0] = 1.
+        state_probs_nox = np.zeros(shape=len(self.state_list), dtype=float)
+        state_probs_nox[0] = 1.
+        state_probs_midext = np.zeros(shape=len(self.midexstate_list), dtype=float)
+        state_probs_midext[0] = 1.
+
+        llh_ex = 0. if log else 1.
+        llh_nox = 0. if log else 1.
+
+        #if 1 1 in diagnose matrix midline extension:
+        #state_probs_midext[0] * llh_nox + state_probs_midext[1] * llh_ex
+
+        for i in range(max_t):
+            for stage in t_stages:
+                state_probs_nox["ipsi"] = self.noext.ipsi._evolve_onestep(start_state = state_probs_nox["ipsi"])
+                state_probs_nox["contra"] = self.noext.contra._evolve_onestep(start_state = state_probs_nox["contra"])
+                state_probs_ex["ipsi"] = state_probs_midext[1] * self.ext.ipsi._evolve_onestep(start_state = state_probs_ex["ipsi"]) + state_probs_midext[0] * state_probs_nox["ipsi"]
+                state_probs_ex["contra"] = state_probs_midext[1] * self.ext.contra._evolve_onestep(start_state = state_probs_ex["contra"]) + state_probs_midext[0] * state_probs_nox["contra"]
+                state_probs_midext = self._evolve_midext(start_midextstate = state_probs_midext)
+
+                joint_state_probs_nox = (
+                    state_probs_nox["ipsi"].T
+                    @ np.diag(self.ipsi.diag_time_dists[stage].pmf)
+                    @ state_probs_nox["contra"]
+                )
+                joint_state_probs_ex = (
+                    state_probs_ex["ipsi"].T
+                    @ np.diag(self.ipsi.diag_time_dists[stage].pmf)
+                    @ state_probs_ex["contra"]
+                )
+                p_ex = np.sum(
+                    self.ext.ipsi.diagnose_matrices[stage]
+                    * (joint_state_probs_ex
+                        @ self.ext.contra.diagnose_matrices[stage]),
+                    axis=0
+                    )
+                p_nox = np.sum(
+                    self.noext.ipsi.diagnose_matrices[stage]
+                    * (joint_state_probs_nox
+                        @ self.noext.contra.diagnose_matrices[stage]),
+                    axis=0
+                    )
+                if log:
+                    llh_ex += np.sum(np.log(p_ex))
+                    llh_nox += np.sum(np.log(p_nox))
+                else:
+                    llh_ex *= np.prod(p_ex)
+                    llh_nox *= np.prod(p_nox)
+
+        return llh_ex + llh_nox
```

Comment:
There should be a `*` in case of `log=True`, right?

Reply:
yes, of course. Thank you. I have to rework this part of the function because the likelihood of the patients with unknown midline extension and the midline diagnose matrix is missing